### PR TITLE
Feat/r17

### PR DESCRIPTION
**Pull request by HanHongChen**
upgrade openapi to r17

### CODE REVIEW

File: internal/sbi/consumer/nrf_service.go [hunk 1]
```diff
@@ -207,27 +197,28 @@ func (s *nnrfService) SendRegisterNFInstance(ctx context.Context) (
...
-		defer func() {
-			if resCloseErr := res.Body.Close(); resCloseErr != nil {
-				logger.ConsumerLog.Errorf("RegisterNFInstance response body cannot close: %+v", resCloseErr)
-			}
-		}()
-		status := res.StatusCode
-		if status == http.StatusOK {
+		nf = res.NrfNfManagementNfProfile
+
+		// http.StatusOK
```
Comment (yccodr) [inline internal/sbi/consumer/nrf_service.go:215]: Remove unnecessary comment. 

```suggestion
```

File: internal/sbi/processor/ampolicy.go [hunk 2]
```diff
@@ -214,33 +215,34 @@ func (p *Processor) PostPoliciesProcedure(polAssoId string,
...
 	if err != nil {
 		return nil, "", pd
 	}
 
 	if amPolicy == nil || amPolicy.AmPolicyData == nil {
 		client := util.GetNudrClient(udrUri)
-		var response *http.Response
-		amData, response, err := client.DefaultApi.PolicyDataUesUeIdAmDataGet(ctx, ue.Supi)
-		if err != nil || response == nil || response.StatusCode != http.StatusOK {
+		// var response *http.Response
```
Comment (yccodr) [inline internal/sbi/processor/ampolicy.go:229]: Remove commented code

```suggestion
```

File: internal/sbi/consumer/nrf_service.go [hunk 3]
```diff
@@ -75,39 +71,29 @@ func (s *nnrfService) getNFDiscClient(uri string) *Nnrf_NFDiscovery.APIClient {
...
 
-	ctx, _, err := s.consumer.Context().GetTokenCtx(models.ServiceName_NNRF_DISC, models.NfType_NRF)
+	ctx, _, err := s.consumer.Context().GetTokenCtx(models.ServiceName_NNRF_DISC, models.NrfNfManagementNfType_NRF)
 	if err != nil {
 		return nil, err
 	}
-
-	result, res, err := client.NFInstancesStoreApi.SearchNFInstances(ctx, targetNfType, requestNfType, &param)
+	res, err := client.NFInstancesStoreApi.SearchNFInstances(ctx, &param)
+	result := res.SearchResult
```
Comment (andy89923) [inline internal/sbi/consumer/nrf_service.go:85]: check res not nil first!

File: internal/sbi/consumer/nrf_service.go [hunk 4]
```diff
@@ -144,11 +130,11 @@ func (s *nnrfService) SendNFInstancesBSF(nrfUri string) string {
...
-	targetNfType := models.NfType_AMF
-	requestNfType := models.NfType_PCF
+	targetNfType := models.NrfNfManagementNfType_AMF
+	requestNfType := models.NrfNfManagementNfType_PCF
 
-	localVarOptionals := Nnrf_NFDiscovery.SearchNFInstancesParamOpts{
-		Guami: optional.NewInterface(util.MarshToJsonString(guami)),
+	localVarOptionals := NFDiscovery.SearchNFInstancesRequest{
+		Guami: &guami,
 	}
```
Comment (andy89923) [inline internal/sbi/consumer/nrf_service.go:138]: SearchNFInstances failed: TargetNfType must be non nil

File: internal/sbi/consumer/nrf_service.go [hunk 5]
```diff
@@ -125,9 +111,9 @@ func (s *nnrfService) SendNFInstancesUDR(nrfUri, id string) string {
 }
 
 func (s *nnrfService) SendNFInstancesBSF(nrfUri string) string {
-	targetNfType := models.NfType_BSF
-	requestNfType := models.NfType_PCF
-	localVarOptionals := Nnrf_NFDiscovery.SearchNFInstancesParamOpts{}
+	targetNfType := models.NrfNfManagementNfType_BSF
+	requestNfType := models.NrfNfManagementNfType_PCF
+	localVarOptionals := NFDiscovery.SearchNFInstancesRequest{}
```
Comment (andy89923) [inline internal/sbi/consumer/nrf_service.go:116]: SearchNFInstances failed: TargetNfType must be non nil

File: internal/sbi/consumer/nrf_service.go [hunk 6]
```diff
@@ -75,39 +71,29 @@ func (s *nnrfService) getNFDiscClient(uri string) *Nnrf_NFDiscovery.APIClient {
...
 	return &result, nil
 }
 
 func (s *nnrfService) SendNFInstancesUDR(nrfUri, id string) string {
-	targetNfType := models.NfType_UDR
-	requestNfType := models.NfType_PCF
-	localVarOptionals := Nnrf_NFDiscovery.SearchNFInstancesParamOpts{
+	targetNfType := models.NrfNfManagementNfType_UDR
+	requestNfType := models.NrfNfManagementNfType_PCF
+	localVarOptionals := NFDiscovery.SearchNFInstancesRequest{
```
Comment (andy89923) [inline internal/sbi/consumer/nrf_service.go:96]: SearchNFInstances failed: TargetNfType must be non nil

File: internal/sbi/processor/policyauthorization.go [hunk 7]
```diff
@@ -997,19 +1016,22 @@ func (p *Processor) SendAppSessionEventNotification(appSession *pcf_context.AppS
...
 
 		request.EvSubsUri = fmt.Sprintf("%s/events-subscription",
 			util.GetResourceUri(models.ServiceName_NPCF_POLICYAUTHORIZATION, appSession.AppSessionId))
 		client := util.GetNpcfPolicyAuthorizationCallbackClient()
-		httpResponse, err := client.PolicyAuthorizationEventNotificationApi.PolicyAuthorizationEventNotification(
-			ctx, uri, request)
+		req := PolicyAuthorization.UpdateEventsSubscEventNotificationPostRequest{
+			PcfPolicyAuthorizationEventsNotification: &request,
+		}
+		httpResponse, err := client.EventsSubscriptionDocumentApi.UpdateEventsSubscEventNotificationPost(
```
[the remark this answers is not in the document]
Comment (HanHongChen) [reply internal/sbi/processor/policyauthorization.go:1030]: The later one. Fixed.

File: internal/sbi/processor/ampolicy.go [hunk 8]
```diff
@@ -214,33 +213,30 @@ func (p *Processor) PostPoliciesProcedure(polAssoId string,
...
 			return nil, "", &problemDetail
 		}
-		defer func() {
-			if rspCloseErr := response.Body.Close(); rspCloseErr != nil {
-				logger.AmPolicyLog.Errorf("PolicyDataUesUeIdAmDataGet response cannot close: %+v", rspCloseErr)
-			}
-		}()
-		if amPolicy == nil {
+
+		if response == nil {
```
[the remark this answers is not in the document]
Comment (andy89923) [reply internal/sbi/processor/ampolicy.go:233]: which is correct?

File: internal/context/ue.go [hunk 9]
```diff
@@ -119,7 +122,7 @@ func (ue *UeContext) NewUeAMPolicyData(assolId string, req models.PolicyAssociat
 		Rfsp:              req.Rfsp,
 		Guami:             req.Guami,
 		UserLoc:           req.UserLoc,
-		ServiveName:       req.ServiveName,
+		ServiceName:       req.ServiceName,
```
Comment (KunLee76) [inline internal/context/ue.go:125]: According to Newest openapi hash, you'll meet this error: `ServiceName: req.ServiveName` for type PcfAmPolicyControlPolicyAssociationRequest struct
So maybe you can create a new PR to fix the new OpenAPI in models/model_pcf_am_policy_control_policy_association_request.go

File: go.mod [hunk 10]
```diff
@@ -3,10 +3,10 @@ module github.com/free5gc/pcf
 go 1.21
 
 require (
-	github.com/antihax/optional v1.0.0
 	github.com/asaskevich/govalidator v0.0.0-20210307081110-f21760c49a8d
 	github.com/cydev/zero v0.0.0-20160322155811-4a4535dd56e7
-	github.com/free5gc/openapi v1.0.8
+	// github.com/free5gc/openapi v1.0.9-0.20240823100026-6a9f667e914f
+	github.com/free5gc/openapi v1.0.9-0.20241024063921-c9ff412a31a6
```
Comment (KunLee76) [inline go.mod:9]: Newest openpai hash: 
`github.com/free5gc/openapi v1.0.9-0.20241024114001-1414ef920de2`
And then you should notice the ue.go in context. (I left a comment)

File: internal/sbi/api_policyauthorization.go [hunk 11]
```diff
@@ -23,38 +23,55 @@ import (
...
+			Method:  http.MethodPost,
+			Pattern: "/app-sessions/pcscf-restoration",
+			APIFunc: s.HTTPPcscfRestoration,
+		},
 	}
 }
 
+// PcscfRestoration - Indicates P-CSCF restoration and does not create an Individual Application Session Context
+func (s *Server) HTTPPcscfRestoration(c *gin.Context) {
+	c.JSON(http.StatusNotImplemented, nil)
```
Comment (KunLee76) [inline internal/sbi/api_policyauthorization.go:72]: Maybe changing the second parameter of c.JSON to `gin.H{}` would be better.
`gin.H{}` returns an empty JSON object ({}). As same as the others' changing way (refer to AMF).

File: internal/sbi/api_uepolicy.go [hunk 12]
```diff
@@ -18,44 +18,44 @@ import (
...
-			APIFunc: s.PoliciesPost,
+			APIFunc: s.HTTPCreateIndividualUEPolicyAssociation,
 		},
 	}
 }
 
-// PoliciesPolAssoIdDelete -
-func (s *Server) PoliciesPolAssoIdDelete(c *gin.Context) {
+func (s *Server) HTTPDeleteIndividualUEPolicyAssociation(c *gin.Context) {
 	c.JSON(http.StatusNotImplemented, nil)
```
Comment (KunLee76) [inline internal/sbi/api_uepolicy.go:48]: `nil` to `gin.H{}`

File: internal/sbi/api_uepolicy.go [hunk 13]
```diff
@@ -18,44 +18,44 @@ import (
...
 }
 
-// PoliciesPolAssoIdGet -
-func (s *Server) PoliciesPolAssoIdGet(c *gin.Context) {
+func (s *Server) HTTPReadIndividualUEPolicyAssociation(c *gin.Context) {
 	c.JSON(http.StatusNotImplemented, nil)
```
Comment (KunLee76) [inline internal/sbi/api_uepolicy.go:52]: `nil` to `gin.H{}`

File: internal/sbi/api_uepolicy.go [hunk 14]
```diff
@@ -18,44 +18,44 @@ import (
...
 }
 
-// PoliciesPolAssoIdUpdatePost -
-func (s *Server) PoliciesPolAssoIdUpdatePost(c *gin.Context) {
+func (s *Server) HTTPReportObservedEventTriggersForIndividualUEPolicyAssociation(c *gin.Context) {
 	c.JSON(http.StatusNotImplemented, nil)
```
Comment (KunLee76) [inline internal/sbi/api_uepolicy.go:56]: `nil` to `gin.H{}`

File: internal/sbi/api_uepolicy.go [hunk 15]
```diff
@@ -18,44 +18,44 @@ import (
...
 }
 
-// PoliciesPost -
-func (s *Server) PoliciesPost(c *gin.Context) {
+func (s *Server) HTTPCreateIndividualUEPolicyAssociation(c *gin.Context) {
 	c.JSON(http.StatusNotImplemented, nil)
```
Comment (KunLee76) [inline internal/sbi/api_uepolicy.go:60]: `nil` to `gin.H{}`

**Review by KunLee76 — CHANGES_REQUESTED** [review]

File: internal/context/ue.go (hunk 9 above)
Comment (HanHongChen) [reply internal/context/ue.go:125]: image
I found out that OpenAPI file is correct.

File: internal/sbi/consumer/udr_service.go [hunk 16]
```diff
@@ -43,6 +43,183 @@ func (s *nudrService) getDataSubscription(uri string) *Nudr_DataRepository.APICl
...
+	param := DataRepository.ReadSessionManagementPolicyDataRequest{
+		UeId:   &ueId,
+		Snssai: sliceInfo,
+		Dnn:    &dnn,
+	}
+	resp, err = client.SessionManagementPolicyDataDocumentApi.ReadSessionManagementPolicyData(ctx, &param)
+	if err == nil {
+		return resp, nil, nil
+	}
+	problem := err.(openapi.GenericOpenAPIError).Model().(models.ProblemDetails)
```
Comment (andy89923) [inline internal/sbi/consumer/udr_service.go:92]: You can't do this. 
**Must use type assertion!** 

There are many place use this, please modify it!

File: internal/sbi/processor/ampolicy.go [hunk 17]
```diff
@@ -308,8 +300,8 @@ func (p *Processor) PostPoliciesProcedure(polAssoId string,
 }
 
 // Send AM Policy Update to AMF if policy has changed
-func (p *Processor) SendAMPolicyUpdateNotification(ue *pcf_context.UeContext,
-	PolId string, request models.PolicyUpdate,
+func (p *Processor) AMPolicyUpdateNotification(ue *pcf_context.UeContext,
```
Comment (andy89923) [inline internal/sbi/processor/ampolicy.go:303]: Hi @HanHongChen ,

The logic here should move to the consumer `SendAMPolicyUpdateNotification()`. 
We will not check the notifier status and result here.  (Should be in `Consumer`)

Comment (HanHongChen) [reply internal/sbi/processor/ampolicy.go]: fixed

**Review by ianchen0119 — APPROVED** [review]

**Pull request merged**